File: apps/backend/app/schemas/analysis.py
"""Analiz job şemaları — ADR-0001 §5 "Aşama B" ve §6.

Frontend aynası: `apps/web/src/lib/api/schemas/analysis.ts`.
"""

from enum import StrEnum
from typing import Literal, Self
from uuid import UUID

from pydantic import Field, field_validator, model_validator

from app.schemas.base import ApiModel, ApiRequestModel, UtcDateTime
from app.schemas.common import ProblemDetails


class AnalysisStatus(StrEnum):
    """queued -> validating -> preprocessing -> analyzing -> aggregating -> completed

    Terminal: completed, failed, cancelled.
    """

    QUEUED = "queued"
    VALIDATING = "validating"
    PREPROCESSING = "preprocessing"
    ANALYZING = "analyzing"
    AGGREGATING = "aggregating"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


TERMINAL_ANALYSIS_STATUSES: frozenset[AnalysisStatus] = frozenset(
    {AnalysisStatus.COMPLETED, AnalysisStatus.FAILED, AnalysisStatus.CANCELLED}
)

# Worker ilerlemesi aşama sınırlarında bu tek tablodan yazılır.
STAGE_PROGRESS: dict[AnalysisStatus, float] = {
    AnalysisStatus.QUEUED: 0.0,
    AnalysisStatus.VALIDATING: 5.0,
    AnalysisStatus.PREPROCESSING: 20.0,
    AnalysisStatus.ANALYZING: 40.0,
    AnalysisStatus.AGGREGATING: 90.0,
    AnalysisStatus.COMPLETED: 100.0,
    AnalysisStatus.FAILED: 100.0,
    AnalysisStatus.CANCELLED: 100.0,
}

# Faz 3/4 kodunun eski adları; public şema adları yukarıdaki dondurulmuş
# sözleşmede kalır.
TERMINAL_STATUSES = TERMINAL_ANALYSIS_STATUSES


class ExportFormat(StrEnum):
    XLSX = "xlsx"
    JSON = "json"


class ModelId(StrEnum):
    """BE-01'de dondurulmuş, structured-output destekli model whitelist'i."""

    CLAUDE_SONNET_4_6 = "anthropic/claude-sonnet-4.6"
    GPT_4_1_MINI = "openai/gpt-4.1-mini"
    GEMINI_2_5_FLASH = "google/gemini-2.5-flash"


class PromptVersion(StrEnum):
    """Backend'de sürümlenmiş ve dağıtıma dâhil prompt'lar."""

    FAQ_ANALYSIS_V1 = "faq_analysis/v1"
    FAQ_ANALYSIS_V2 = "faq_analysis/v2"
    FAQ_ANALYSIS_V3 = "faq_analysis/v3"
    FAQ_ANALYSIS_V4 = "faq_analysis/v4"


class AnalysisMode(StrEnum):
    """Bir LLM kaydının hangi kaynak birimini temsil ettiği."""

    MESSAGE = "message"
    CONTEXTUAL_USER_TURNS = "contextual_user_turns"


class ConversationConfig(ApiRequestModel):
    """Session bağlamlı kullanıcı-turn analizi için kolon/değer eşlemesi.

    Kullanıcı turn'ü sınıflandırılan ve sayılan birimdir. Aynı session'daki
    önceki kullanıcı mesajları bağlam olabilir. Bot mesajları varsayılan
    olarak elenir; yalnızca açıkça istendiğinde bağlam olur ve hiçbir zaman
    rapor adetlerine girmez. Rol ve mesaj türü değerleri tam eşleşmeyle
    yorumlanır.
    """

    session_id_column: str = Field(min_length=1, max_length=512)
    message_order_column: str = Field(min_length=1, max_length=512)
    role_column: str = Field(min_length=1, max_length=512)
    message_type_column: str = Field(min_length=1, max_length=512)
    user_role_values: list[str] = Field(
        default_factory=lambda: ["Kullanıcı"], min_length=1, max_length=20
    )
    assistant_role_values: list[str] = Field(
        default_factory=lambda: ["Bot"], min_length=1, max_length=20
    )
    include_assistant_context: bool = False
    target_message_types: list[str] = Field(
        # Quick-reply tıklamaları niyeti anlamada bağlam olabilir, fakat
        # varsayılan SSS paydasını kullanıcı buton cevaplarıyla şişirmemeli.
        default_factory=lambda: ["text"],
        min_length=1,
        max_length=20,
    )
    context_message_types: list[str] = Field(
        default_factory=lambda: ["text", "quick_reply", "single-choice"],
        min_length=1,
        max_length=20,
    )
    max_context_turns: int = Field(default=4, ge=1, le=8)
    max_context_tokens: int = Field(default=1000, ge=128, le=4000)

    @field_validator(
        "session_id_column",
        "message_order_column",
        "role_column",
        "message_type_column",
    )
    @classmethod
    def _strip_mapping_column(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("Konuşma eşleme kolonu boş olamaz.")
        return stripped

    @field_validator(
        "user_role_values",
        "assistant_role_values",
        "target_message_types",
        "context_message_types",
    )
    @classmethod
    def _normalize_mapping_values(cls, values: list[str]) -> list[str]:
        normalized = [value.strip() for value in values]
        if any(not value for value in normalized):
            raise ValueError("Konuşma eşleme değerleri boş olamaz.")
        if any(len(value) > 512 for value in normalized):
            raise ValueError("Konuşma eşleme değerleri en fazla 512 karakter olabilir.")
        if len(normalized) != len(set(normalized)):
            raise ValueError("Konuşma eşleme değerleri tekrarlanamaz.")
        return normalized

    @model_validator(mode="after")
    def _mapping_is_unambiguous(self) -> Self:
        columns = [
            self.session_id_column,
            self.message_order_column,
            self.role_column,
            self.message_type_column,
        ]
        if len(columns) != len(set(columns)):
            raise ValueError("Session, sıra, rol ve mesaj türü kolonları farklı olmalı.")
        if set(self.user_role_values) & set(self.assistant_role_values):
            raise ValueError("Aynı rol değeri hem kullanıcı hem bot olamaz.")
        if not set(self.target_message_types) <= set(self.context_message_types):
            raise ValueError("Hedef mesaj türleri bağlam mesaj türlerinin alt kümesi olmalı.")
        return self


class RowFilter(ApiRequestModel):
    """Bir kolonda izin verilen değerler — filtreler arası AND, değerler arası OR.

    Filtre karşılaştırması XLSX hücresinin metne çevrilmiş ve kenar boşlukları
    temizlenmiş hâliyle TAM eşleşmedir. Regex/substring bilinçli olarak yok:
    kullanıcı verisine karşı tahmin edilemez veya pahalı bir ifade motoru
    çalıştırmıyoruz.
    """

    column: str = Field(min_length=1, max_length=512)
    allowed_values: list[str] = Field(min_length=1, max_length=20)

    @field_validator("column")
    @classmethod
    def _strip_column(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("Filtre kolonu boş olamaz.")
        return stripped

    @field_validator("allowed_values")
    @classmethod
    def _normalize_allowed_values(cls, values: list[str]) -> list[str]:
        normalized = [value.strip() for value in values]
        if any(not value for value in normalized):
            raise ValueError("Filtre değerleri boş olamaz.")
        if any(len(value) > 512 for value in normalized):
            raise ValueError("Filtre değerleri en fazla 512 karakter olabilir.")
        if len(normalized) != len(set(normalized)):
            raise ValueError("Aynı filtre değeri tekrarlanamaz.")
        return normalized


class AnalysisRequest(ApiRequestModel):
    """POST /api/v1/analyses gövdesi.

    OpenRouter anahtarı BU GÖVDEDE YER ALMAZ — yalnızca `X-OpenRouter-Key`
    header'ında taşınır (ADR-0001 §6/§9). `RequestValidationError` handler'ı
    `errors[].input` alanını asla yankılamaz; anahtar yanlışlıkla gövdeye
    konursa geri sızdırılmaması için (ADR-0002 #7).
    """

    upload_id: UUID
    sheet_name: str = Field(min_length=1)
    text_column: str = Field(min_length=1)
    row_filters: list[RowFilter] = Field(default_factory=list, max_length=5)
    analysis_mode: AnalysisMode = AnalysisMode.MESSAGE
    conversation_config: ConversationConfig | None = None
    model: ModelId
    prompt_version: PromptVersion
    top_n: int = Field(ge=1, le=100)
    #: gt, ge değil: Zod tarafı `.positive()`.
    max_cost_usd: float = Field(gt=0, le=100)

    @field_validator("row_filters")
    @classmethod
    def _filter_columns_are_unique(cls, filters: list[RowFilter]) -> list[RowFilter]:
        columns = [row_filter.column for row_filter in filters]
        if len(columns) != len(set(columns)):
            raise ValueError("Aynı kolon için birden fazla filtre tanımlanamaz.")
        return filters

    @model_validator(mode="after")
    def _mode_configuration_is_consistent(self) -> Self:
        if self.analysis_mode is AnalysisMode.MESSAGE:
            if self.conversation_config is not None:
                raise ValueError(
                    "conversation_config yalnızca contextual_user_turns modunda kullanılabilir."
                )
            if self.prompt_version is PromptVersion.FAQ_ANALYSIS_V4:
                raise ValueError("faq_analysis/v4 yalnızca contextual_user_turns içindir.")
            return self

        config = self.conversation_config
        if config is None:
            raise ValueError("contextual_user_turns modu conversation_config gerektirir.")
        if self.prompt_version is not PromptVersion.FAQ_ANALYSIS_V4:
            raise ValueError("contextual_user_turns modu faq_analysis/v4 prompt'unu gerektirir.")

        mapped_columns = {
            config.session_id_column,
            config.message_order_column,
            config.role_column,
            config.message_type_column,
        }
        if self.text_column in mapped_columns:
            raise ValueError("Metin kolonu konuşma eşleme kolonlarından farklı olmalı.")
        if mapped_columns & {row_filter.column for row_filter in self.row_filters}:
            raise ValueError(
                "Session, sıra, rol ve mesaj türü kolonları genel satır filtresi olarak "
                "kullanılamaz. Bu değerleri konuşma ayarlarından eşleyin."
            )
        return self


class AnalysisCreated(ApiModel):
    """POST /api/v1/analyses — 202 Accepted."""

    analysis_id: UUID
    status: Literal[AnalysisStatus.QUEUED] = AnalysisStatus.QUEUED


class AnalysisJob(ApiModel):
    """GET /api/v1/analyses/{analysis_id} — durum sorgulaması.

    ADR-0001 §2: ilerleme her satırda değil, yalnızca aşama veya anlamlı yüzde
    değişiminde yazılır; `progress` düzenli artmayabilir.
    """

    analysis_id: UUID
    status: AnalysisStatus
    progress: float = Field(ge=0, le=100)
    created_at: UtcDateTime
    updated_at: UtcDateTime
    #: Terminal olmayan durumlarda kalan süre tahmini, saniye.
    estimated_seconds_remaining: float | None = Field(default=None, ge=0)
    #: Yalnızca status "failed" iken dolu.
    error: ProblemDetails | None = None

    @model_validator(mode="after")
    def _payload_matches_status(self) -> Self:
        if self.status is AnalysisStatus.FAILED:
            if self.error is None:
                raise ValueError("failed analysis job error taşımak zorunda.")
        elif self.error is not None:
            raise ValueError("Yalnızca failed analysis job error taşıyabilir.")

        if (
            self.status in TERMINAL_ANALYSIS_STATUSES
            and self.estimated_seconds_remaining is not None
        ):
            raise ValueError("Terminal analysis job kalan süre tahmini taşıyamaz.")
        return self


class ModelOption(ApiModel):
    """GET /api/v1/models öğesi.

    ADR-0001 §6 "model yalnızca backend whitelist'inden seçilebilir" diyordu
    ama listeyi döndüren endpoint'i tanımlamıyordu. ADR-0002 #1 ile
    `GET /api/v1/models` sözleşmeye dâhil edildi.
    """

    id: ModelId
    label: str
    #: 1M girdi tokenı başına USD.
    input_cost_per_million: float = Field(ge=0)
    output_cost_per_million: float = Field(ge=0)
    cache_read_cost_per_million: float | None = Field(default=None, ge=0)
    cache_write_cost_per_million: float | None = Field(default=None, ge=0)
    context_window: int = Field(gt=0)
    pricing_source: Literal["openrouter", "fallback"] = "fallback"
    pricing_updated_at: UtcDateTime | None = None


class PricingSnapshot(ApiModel):
    """Bir job oluşturulurken sabitlenen model fiyatları."""

    input_cost_per_million: float = Field(ge=0)
    output_cost_per_million: float = Field(ge=0)
    cache_read_cost_per_million: float | None = Field(default=None, ge=0)
    cache_write_cost_per_million: float | None = Field(default=None, ge=0)
    source: Literal["openrouter", "fallback"]
    fetched_at: UtcDateTime | None = None


class ModelList(ApiModel):
    """GET /api/v1/models"""

    models: list[ModelOption]
    default_model: ModelId
    default_prompt_version: PromptVersion

    @model_validator(mode="after")
    def _defaults_are_available(self) -> Self:
        ids = [model.id for model in self.models]
        if len(ids) != len(set(ids)):
            raise ValueError("Model whitelist'i tekrarlı id taşıyamaz.")
        if set(ids) != set(ModelId):
            raise ValueError("Model whitelist'i dondurulmuş ModelId kümesiyle aynı olmalı.")
        if self.default_model not in ids:
            raise ValueError("default_model whitelist içinde bulunmalı.")
        return self


AnalysisCreate = AnalysisRequest
AnalysisJobRead = AnalysisJob

File: apps/web/src/lib/api/schemas/analysis.ts
import * as z from "zod";

import { problemDetailsSchema } from "./common";

/**
 * Analiz job'ı şemaları — ADR §6 "Analysis" ve §5 "Aşama B".
 */

/**
 * ADR §6 job durumları:
 *   queued -> validating/preprocessing -> analyzing -> aggregating -> completed
 * Terminal durumlar: failed, cancelled
 */
export const analysisStatusSchema = z.enum([
  "queued",
  "validating",
  "preprocessing",
  "analyzing",
  "aggregating",
  "completed",
  "failed",
  "cancelled",
]);

export type AnalysisStatus = z.infer<typeof analysisStatusSchema>;

/** İlerleme ekranında gösterilecek Türkçe aşama adları. */
export const ANALYSIS_STAGE_LABELS_TR: Record<AnalysisStatus, string> = {
  queued: "Sırada bekliyor",
  validating: "Dosya doğrulanıyor",
  preprocessing: "Veri temizleniyor",
  analyzing: "Mesajlar analiz ediliyor",
  aggregating: "Sonuçlar birleştiriliyor",
  completed: "Tamamlandı",
  failed: "Başarısız",
  cancelled: "İptal edildi",
};

/** Sırası ilerleme göstergesindeki adım sırasını belirler. */
export const ANALYSIS_ACTIVE_STAGES = [
  "queued",
  "validating",
  "preprocessing",
  "analyzing",
  "aggregating",
] as const satisfies readonly AnalysisStatus[];

const TERMINAL_STATUSES: readonly AnalysisStatus[] = ["completed", "failed", "cancelled"];

/**
 * Terminal durumlar polling'i durdurur. TanStack Query'nin refetchInterval'ı
 * bu fonksiyona bakar; yanlış olursa 45 dakikalık bir job bittikten sonra da
 * sonsuza kadar istek atmaya devam eder.
 */
export function isAnalysisSettled(status: AnalysisStatus): boolean {
  return TERMINAL_STATUSES.includes(status);
}

/**
 * BE-01'de dondurulan OpenRouter model whitelist'i.
 *
 * Bu değerler request, model kataloğu ve rapor şemalarında aynı anda
 * kullanılır. Böylece backend'in desteklemediği bir model tel üstünde
 * "herhangi bir string" gibi görünemez.
 */
export const modelIdSchema = z.enum([
  "anthropic/claude-sonnet-4.6",
  "openai/gpt-4.1-mini",
  "google/gemini-2.5-flash",
]);

export type ModelId = z.infer<typeof modelIdSchema>;

/** BE-01'de dondurulan, sürümlenmiş prompt whitelist'i. */
export const promptVersionSchema = z.enum([
  "faq_analysis/v1",
  "faq_analysis/v2",
  "faq_analysis/v3",
  "faq_analysis/v4",
]);

export type PromptVersion = z.infer<typeof promptVersionSchema>;

export const analysisModeSchema = z.enum(["message", "contextual_user_turns"]);

export type AnalysisMode = z.infer<typeof analysisModeSchema>;

const conversationMappingValuesSchema = z
  .array(z.string().trim().min(1, "Eşleme değeri boş olamaz.").max(512))
  .min(1, "En az bir eşleme değeri girilmelidir.")
  .max(20, "En fazla 20 eşleme değeri girilebilir.")
  .superRefine((values, ctx) => {
    if (new Set(values).size !== values.length) {
      ctx.addIssue({ code: "custom", message: "Aynı eşleme değeri tekrarlanamaz." });
    }
  });

export const conversationConfigSchema = z
  .object({
    session_id_column: z.string().trim().min(1, "Oturum kimliği kolonu seçilmelidir.").max(512),
    message_order_column: z.string().trim().min(1, "Mesaj sırası kolonu seçilmelidir.").max(512),
    role_column: z.string().trim().min(1, "Gönderen rolü kolonu seçilmelidir.").max(512),
    message_type_column: z.string().trim().min(1, "Mesaj türü kolonu seçilmelidir.").max(512),
    user_role_values: conversationMappingValuesSchema.default(["Kullanıcı"]),
    assistant_role_values: conversationMappingValuesSchema.default(["Bot"]),
    include_assistant_context: z.boolean().default(false),
    // Quick reply varsayılan olarak yalnız bağlamdır; kullanıcı isterse
    // analiz hedeflerine açıkça ekleyebilir.
    target_message_types: conversationMappingValuesSchema.default(["text"]),
    context_message_types: conversationMappingValuesSchema.default([
      "text",
      "quick_reply",
      "single-choice",
    ]),
    max_context_turns: z
      .int("Bağlam turu sayısı tam sayı olmalıdır.")
      .min(1, "En az 1 önceki tur kullanılmalıdır.")
      .max(8, "En fazla 8 önceki tur kullanılabilir.")
      .default(4),
    max_context_tokens: z
      .int("Bağlam token sınırı tam sayı olmalıdır.")
      .min(128, "Bağlam token sınırı en az 128 olmalıdır.")
      .max(4000, "Bağlam token sınırı en fazla 4000 olabilir.")
      .default(1000),
  })
  .superRefine((config, ctx) => {
    const columns = [
      config.session_id_column,
      config.message_order_column,
      config.role_column,
      config.message_type_column,
    ];
    if (new Set(columns).size !== columns.length) {
      ctx.addIssue({
        code: "custom",
        path: ["message_type_column"],
        message: "Oturum, sıra, rol ve mesaj türü kolonları farklı olmalıdır.",
      });
    }

    const userRoles = new Set(config.user_role_values);
    if (config.assistant_role_values.some((value) => userRoles.has(value))) {
      ctx.addIssue({
        code: "custom",
        path: ["assistant_role_values"],
        message: "Aynı rol değeri hem kullanıcı hem bot olamaz.",
      });
    }

    const contextTypes = new Set(config.context_message_types);
    if (config.target_message_types.some((value) => !contextTypes.has(value))) {
      ctx.addIssue({
        code: "custom",
        path: ["target_message_types"],
        message: "Analiz edilecek mesaj türleri bağlam mesaj türlerinin alt kümesi olmalıdır.",
      });
    }
  });

export type ConversationConfig = z.infer<typeof conversationConfigSchema>;

export const rowFilterSchema = z.object({
  column: z.string().trim().min(1, "Filtre kolonu boş olamaz.").max(512),
  allowed_values: z
    .array(z.string().trim().min(1, "Filtre değeri boş olamaz.").max(512))
    .min(1, "En az bir filtre değeri girilmelidir.")
    .max(20, "Bir filtrede en fazla 20 değer olabilir.")
    .superRefine((values, ctx) => {
      if (new Set(values).size !== values.length) {
        ctx.addIssue({ code: "custom", message: "Aynı filtre değeri tekrarlanamaz." });
      }
    }),
});

export type RowFilter = z.infer<typeof rowFilterSchema>;

/**
 * POST /api/v1/analyses gövdesi.
 *
 * OpenRouter anahtarı BU GÖVDEDE YER ALMAZ. ADR §6/§9 gereği yalnızca
 * X-OpenRouter-Key header'ında taşınır, loglarda redakte edilir ve
 * PostgreSQL'e hiçbir şekilde yazılmaz.
 */
const analysisRequestFieldsSchema = z.object({
  upload_id: z.uuid(),
  sheet_name: z.string().min(1, "Sayfa seçilmelidir."),
  text_column: z.string().min(1, "Analiz edilecek metin kolonu seçilmelidir."),
  row_filters: z
    .array(rowFilterSchema)
    .max(5, "En fazla 5 satır filtresi tanımlanabilir.")
    .superRefine((filters, ctx) => {
      const columns = filters.map((rowFilter) => rowFilter.column);
      if (new Set(columns).size !== columns.length) {
        ctx.addIssue({ code: "custom", message: "Aynı kolon birden fazla filtrelenemez." });
      }
    }),
  analysis_mode: analysisModeSchema.default("message"),
  conversation_config: conversationConfigSchema.nullable().default(null),
  model: modelIdSchema,
  prompt_version: promptVersionSchema,
  top_n: z.int().min(1, "En az 1 sonuç istenmelidir.").max(100, "En fazla 100 sonuç istenebilir."),
  max_cost_usd: z
    .number()
    .positive("Maliyet sınırı sıfırdan büyük olmalıdır.")
    .max(100, "Maliyet sınırı en fazla 100 USD olabilir."),
});

type AnalysisConfigurationCandidate = Omit<
  z.infer<typeof analysisRequestFieldsSchema>,
  "upload_id"
>;

function validateAnalysisConfiguration(
  request: AnalysisConfigurationCandidate,
  ctx: z.RefinementCtx,
) {
  if (request.analysis_mode === "message") {
    if (request.conversation_config !== null) {
      ctx.addIssue({
        code: "custom",
        path: ["conversation_config"],
        message: "Konuşma ayarları yalnızca bağlamsal kullanıcı turları modunda kullanılabilir.",
      });
    }
    if (request.prompt_version === "faq_analysis/v4") {
      ctx.addIssue({
        code: "custom",
        path: ["prompt_version"],
        message: "faq_analysis/v4 yalnızca bağlamsal kullanıcı turları içindir.",
      });
    }
    return;
  }

  const config = request.conversation_config;
  if (config === null) {
    ctx.addIssue({
      code: "custom",
      path: ["conversation_config"],
      message: "Bağlamsal kullanıcı turları modu konuşma ayarlarını gerektirir.",
    });
    return;
  }

  if (request.prompt_version !== "faq_analysis/v4") {
    ctx.addIssue({
      code: "custom",
      path: ["prompt_version"],
      message: "Bağlamsal kullanıcı turları modu faq_analysis/v4 kullanmalıdır.",
    });
  }

  const mappedColumns = [
    ["session_id_column", config.session_id_column],
    ["message_order_column", config.message_order_column],
    ["role_column", config.role_column],
    ["message_type_column", config.message_type_column],
  ] as const;
  for (const [field, column] of mappedColumns) {
    if (column === request.text_column) {
      ctx.addIssue({
        code: "custom",
        path: ["conversation_config", field],
        message: "Metin kolonu konuşma eşleme kolonlarından farklı olmalıdır.",
      });
    }
  }

  const filterColumns = new Set(request.row_filters.map((rowFilter) => rowFilter.column));
  if (mappedColumns.some(([, column]) => filterColumns.has(column))) {
    ctx.addIssue({
      code: "custom",
      path: ["row_filters"],
      message: "Konuşma eşleme kolonları genel satır filtresi olarak kullanılamaz.",
    });
  }
}

export const analysisRequestSchema = analysisRequestFieldsSchema.superRefine(
  validateAnalysisConfiguration,
);

/** Formun URL'den aldığı upload_id dışındaki analiz ayarları. */
export const analysisConfigurationSchema = analysisRequestFieldsSchema
  .omit({ upload_id: true })
  .superRefine(validateAnalysisConfiguration);

/** İstemciler backend defaultlarını kullanmak için yeni alanları atlayabilir. */
export type AnalysisRequest = z.input<typeof analysisRequestSchema>;

/** Doğrulama sonrası tüm defaultları uygulanmış, fingerprint'e hazır istek. */
export type ValidatedAnalysisRequest = z.output<typeof analysisRequestSchema>;

/** POST /api/v1/analyses — 202 Accepted cevabı. */
export const analysisCreatedSchema = z.object({
  analysis_id: z.uuid(),
  status: z.literal("queued"),
});

export type AnalysisCreated = z.infer<typeof analysisCreatedSchema>;

/**
 * GET /api/v1/analyses/{analysis_id} cevabı — durum sorgulaması.
 *
 * ADR §2: ilerleme her satırda değil, yalnızca aşama veya anlamlı yüzde
 * değişiminde yazılır. Yani `progress` düzenli artmayabilir; arayüz
 * bunu düz bir sayaç gibi göstermemelidir.
 */
export const analysisJobSchema = z
  .object({
    analysis_id: z.uuid(),
    status: analysisStatusSchema,
    /** 0-100 arası tamamlanma yüzdesi. */
    progress: z.number().min(0).max(100),
    created_at: z.iso.datetime(),
    updated_at: z.iso.datetime(),
    /** Terminal olmayan durumlarda backend'in kalan süre tahmini, saniye. */
    estimated_seconds_remaining: z.number().nonnegative().nullable().default(null),
    /** Yalnızca status "failed" iken dolu. */
    error: problemDetailsSchema.nullable().default(null),
  })
  .superRefine((job, ctx) => {
    if ((job.status === "failed") !== (job.error !== null)) {
      ctx.addIssue({
        code: "custom",
        path: ["error"],
        message: "Job error alanı status ile uyumsuz.",
      });
    }
    if (TERMINAL_STATUSES.includes(job.status) && job.estimated_seconds_remaining !== null) {
      ctx.addIssue({
        code: "custom",
        path: ["estimated_seconds_remaining"],
        message: "Terminal job kalan süre tahmini taşıyamaz.",
      });
    }
  });

export type AnalysisJob = z.infer<typeof analysisJobSchema>;

/**
 * Model seçim listesi öğesi — `GET /api/v1/models`.
 *
 * ADR-0001 §6 "model yalnızca backend whitelist'inden seçilebilir" diyordu
 * ama bu listeyi döndüren bir endpoint tanımlamıyordu. ADR-0002 #1 ile uç
 * sözleşmeye dâhil edildi; artık kesinleşmiştir ve `docs/api/openapi.json`
 * üzerinden doğrulanır.
 */
export const modelOptionSchema = z.object({
  id: modelIdSchema,
  label: z.string(),
  /** 1M girdi tokenı başına USD; maliyet tahmini için. */
  input_cost_per_million: z.number().nonnegative(),
  output_cost_per_million: z.number().nonnegative(),
  cache_read_cost_per_million: z.number().nonnegative().nullable().default(null),
  cache_write_cost_per_million: z.number().nonnegative().nullable().default(null),
  context_window: z.int().positive(),
  pricing_source: z.enum(["openrouter", "fallback"]).default("fallback"),
  pricing_updated_at: z.iso.datetime().nullable().default(null),
});

export type ModelOption = z.infer<typeof modelOptionSchema>;

export const pricingSnapshotSchema = z.object({
  input_cost_per_million: z.number().nonnegative(),
  output_cost_per_million: z.number().nonnegative(),
  cache_read_cost_per_million: z.number().nonnegative().nullable().default(null),
  cache_write_cost_per_million: z.number().nonnegative().nullable().default(null),
  source: z.enum(["openrouter", "fallback"]),
  fetched_at: z.iso.datetime().nullable().default(null),
});

export type PricingSnapshot = z.infer<typeof pricingSnapshotSchema>;

export const modelListSchema = z
  .object({
    models: z.array(modelOptionSchema),
    default_model: modelIdSchema,
    default_prompt_version: promptVersionSchema,
  })
  .superRefine((catalog, ctx) => {
    const ids = catalog.models.map((model) => model.id);
    if (new Set(ids).size !== ids.length) {
      ctx.addIssue({
        code: "custom",
        path: ["models"],
        message: "Model id'leri benzersiz olmalı.",
      });
    }
    const actualIds = new Set(ids);
    if (
      modelIdSchema.options.some((id) => !actualIds.has(id)) ||
      actualIds.size !== modelIdSchema.options.length
    ) {
      ctx.addIssue({
        code: "custom",
        path: ["models"],
        message: "Model listesi dondurulmuş whitelist ile birebir aynı olmalı.",
      });
    }
    if (!ids.includes(catalog.default_model)) {
      ctx.addIssue({
        code: "custom",
        path: ["default_model"],
        message: "Varsayılan model whitelist içinde olmalı.",
      });
    }
  });

export type ModelList = z.infer<typeof modelListSchema>;

/** GET /api/v1/analyses/{analysis_id}/export?format=xlsx|json */
export const exportFormatSchema = z.enum(["xlsx", "json"]);

export type ExportFormat = z.infer<typeof exportFormatSchema>;
